apps/library/include/gpc_memory_layout.h

```diff
@@ -0,0 +1,136 @@
+#ifndef _GPC_MEMORY_LAYOUT_H_
+#define _GPC_MEMORY_LAYOUT_H_
+
+/**
+ * Default values for memory layout constants. Actual values can be overridden by build configuration
+ * Used to generate LD script and may be included by entry point, so it must contain preprocessor
+ * directives only
+ */
+
+#include "gpc_defs.h"
+
+/**
+ * Number of agent ID bits in GPC memory address
+ */
+#ifndef GPC_MEM_AGENT_ID_BITS
+    #define GPC_MEM_AGENT_ID_BITS (8)
+#endif
+
+// Base of CR space address (CR region at the right place, other bits zero)
+#define GPC_MEM_REGION_OFFSET (32 - 2 - GPC_MEM_AGENT_ID_BITS)
+#define GPC_MEM_REGION_BASE(region) ((region) << GPC_MEM_REGION_OFFSET)
+
+/**
+ * I_MEM size
+ */
+#ifndef GPC_MEM_I_MEM_SIZE
+#define GPC_MEM_I_MEM_SIZE 0x800
+#endif
+
+/**
+ * D_MEM size
+ */
+#ifndef GPC_MEM_D_MEM_SIZE
+#define GPC_MEM_D_MEM_SIZE 0x800
+#endif
+
+/**
+ * Base address of D_MEM
+ */
+#define GPC_MEM_D_MEM_BASE GPC_MEM_REGION_BASE(GPC_MEM_REGION_D_MEM)
+
+/**
+ * Base address of I_MEM
+ */
+#define GPC_MEM_I_MEM_BASE GPC_MEM_REGION_BASE(GPC_MEM_REGION_I_MEM)
+
+/**
+ * Base address of CR space
+ */
+#define GPC_MEM_CR_BASE GPC_MEM_REGION_BASE(GPC_MEM_REGION_CR)
+
+/**
+ * Default D_MEM layout (sizes for 2k, see relative size for division)
+ *
+ * +========+======+===============+===============+
+ * | Offset | Size |    Region     | Relative size |
+ * +========+======+===============+===============+
+ * |   1936 |  112 | Stack[3]      | 7/128         |
+ * +--------+------+---------------+---------------+
+ * |   1824 |  112 | Stack[2]      | 7/128         |
+ * +--------+------+---------------+---------------+
+ * |   1712 |  112 | Stack[1]      | 7/128         |
+ * +--------+------+---------------+---------------+
+ * |   1600 |  112 | Stack[0]      | 7/128         |
+ * +--------+------+---------------+---------------+
+ * |   1264 |  336 | Local data[3] | 21/128        |
+ * +--------+------+---------------+---------------+
+ * |    928 |  336 | Local data[2] | 21/128        |
+ * +--------+------+---------------+---------------+
+ * |    592 |  336 | Local data[1] | 21/128        |
+ * +--------+------+---------------+---------------+
+ * |    256 |  336 | Local data[0] | 21/128        |
+ * +--------+------+---------------+---------------+
+ * |    128 |  256 | Global data   | 1/8 - 128B    |
+ * +--------+------+---------------+---------------+
+ * |     96 |   32 | GPC params[3] |               |
+ * +--------+------+---------------+---------------+
+ * |     64 |   32 | GPC params[2] |               |
+ * +--------+------+---------------+---------------+
+ * |     32 |   32 | GPC params[1] |               |
+ * +--------+------+---------------+---------------+
+ * |      0 |   32 | GPC params[0] |               |
```

Comment:
What are the 4 "params"?
1) GPC_MEM_LOCAL_DATA_BASE , 
2) GPC_MEM_LOCAL_DATA_SIZE, 
3) GPC_MEM_STACK_BASE, 
4) GPC_MEM_STACK_SIZE?

We have a dedicated register in the CR Space - you don't need to use it, but i see that you allocated memory Space for it - so why not just use the CR Space?

Reply:
These are parameters for software, one per thread. They include software-specific parameters, currently that includes current thread ID and information about memory layout. The parameters are initialized by the entry point code and passed as an argument to user code entry points. Currently 5 words are used and the other three are reserved for future use (perhaps I'll use one of them for launch synchronization).

See "SW Stack Arch" for details